system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -63,70 +101,422 @@ class CLibCompiledXpath : public CInterface, public ICompiledXpath
+        //if we're preventing replacing variables we need to handle elsewhere
+        //  and still replace external values when treated as xsdl:variables, but not when treated as xsdl:params
+        if (ret.first->second)
+            xmlXPathFreeObject(ret.first->second);
+        ret.first->second = obj;
+        return true;
+    }
+    xmlXPathObjectPtr getObject(const char *key)
+    {
+        XPathObjectMap::iterator it = variables.find(key);
+        if (it == variables.end())
+            return nullptr;
+        return it->second;
+    }
 };
 
-class CLibXpathContext : public CInterface, public IXpathContext
+typedef std::vector<std::unique_ptr<CLibXpathScope>> XPathScopeVector;
+typedef std::map<std::string, ICompiledXpath*> XPathInputMap;
+
+class XpathContextState
 {
 private:
+    xmlDocPtr doc = nullptr;
+    xmlNodePtr node = nullptr;
+    int contextSize = 0;
+    int proximityPosition = 0;
+public:
+    XpathContextState(xmlXPathContextPtr ctx)
+    {
+        doc = ctx->doc;
+        node = ctx->node;
+        contextSize = ctx->contextSize;
+        proximityPosition = ctx->proximityPosition;
+    }
+
+    void restore(xmlXPathContextPtr ctx)
+    {
+        ctx->doc = doc;
+        ctx->node = node;
+        ctx->contextSize = contextSize;
+        ctx->proximityPosition = proximityPosition;
+    }
+};
+
+typedef std::vector<XpathContextState> XPathContextStateVector;
+
+class CLibXpathContext : public CInterfaceOf<IXpathContext>
+{
+public:
+    XPathInputMap provided;
     xmlDocPtr m_xmlDoc = nullptr;
     xmlXPathContextPtr m_xpathContext = nullptr;
     StringBuffer m_xpath;
     ReadWriteLock m_rwlock;
+    XPathScopeVector scopes;
+    bool strictParameterDeclaration = true;
 
-public:
-    IMPLEMENT_IINTERFACE;
+    //saved state
+    XPathContextStateVector saved;
 
-    CLibXpathContext(const char * xmldoc) //not thread safe
+public:
+    CLibXpathContext(const char * xmldoc, bool _strictParameterDeclaration) : strictParameterDeclaration(_strictParameterDeclaration)
     {
+        beginScope("/");
         setXmlDoc(xmldoc);
+        exsltDateXpathCtxtRegister(m_xpathContext, (xmlChar*)"date");
+        exsltMathXpathCtxtRegister(m_xpathContext, (xmlChar*)"math");
+        exsltSetsXpathCtxtRegister(m_xpathContext, (xmlChar*)"set");
+        exsltStrXpathCtxtRegister(m_xpathContext, (xmlChar*)"str");
     }
+
     ~CLibXpathContext()
     {
+        for (XPathInputMap::iterator it=provided.begin(); it!=provided.end(); ++it)
+            it->second->Release();
         xmlXPathFreeContext(m_xpathContext);
         xmlFreeDoc(m_xmlDoc);
     }
 
+    void pushLocation()
+    {
+        WriteLockBlock wblock(m_rwlock);
+        saved.emplace_back(XpathContextState(m_xpathContext));
+    }
+
+    void setLocation(xmlDocPtr doc, xmlNodePtr node, int contextSize, int proximityPosition)
+    {
+        WriteLockBlock wblock(m_rwlock);
+        m_xpathContext->doc = doc;
```

Comment:
I don't understand the question.  We keep the old info in our location stack.  This mechanism lets us process the results of nodesets that may refer to temporary locations.  Think of the power of us implementing the str:tokenize function.  This is how xslt works.  Unfortunately their implementation of tokenize is xslt specific.  We could implement out own and other similar functions though and it will let us do really cool stuff.  We can refactor how the locations are managed when we investigate the possibilities further.

Reply:
I wouldn't expect the element nodes returned from an xpath evaluation to refer to a document other than the one pointed to by the context that performed the evaluation. If there are nodes that don't refer to the same document, then I guess it's wise but potentially confusing.

Reply:
yep confusing or not that's how libxml2's xpath, and libxslt (which we're to some degree emulating) work.